src/main/scala/is/hail/io/index/IndexReader.scala

```diff
@@ -83,25 +84,155 @@ class IndexReader(hConf: Configuration, path: String, cacheCapacity: Int = 8) ex
+      val node = readLeafNode(offset)
+      val idx = binarySearchUpperBound(node.children.length, key, { i => node.children(i).key })
+      node.firstIndex + idx
+    } else {
+      val node = readInternalNode(offset)
+      val children = node.children
+      val n = children.length
+      val idx = binarySearchUpperBound(n, key, { i => children(i).firstKey })
+      upperBound(key, level - 1, children(idx - 1).indexFileOffset)
+    }
+  }
+
+  private[io] def upperBound(key: Annotation): Long = {
+    if (nKeys == 0 || ordering.lt(key, readInternalNode(metadata.rootOffset).children.head.firstKey))
+      0
+    else
+      upperBound(key, height - 1, metadata.rootOffset)
+  }
+
+  private def getLeafNode(index: Long, level: Int, offset: Long): LeafNode = {
+    if (level == 0) {
+      readLeafNode(offset)
     } else {
       val node = readInternalNode(offset)
       val firstIndex = node.firstIndex
       val nKeysPerChild = math.pow(branchingFactor, level).toLong
       val localIndex = (index - firstIndex) / nKeysPerChild
-      queryByIndex(index, level - 1, node.children(localIndex.toInt).indexFileOffset)
+      getLeafNode(index, level - 1, node.children(localIndex.toInt).indexFileOffset)
+    }
+  }
+
+  private def getLeafNode(index: Long): LeafNode =
+    getLeafNode(index, height - 1, metadata.rootOffset)
+
+  def queryByKey(key: Annotation): Array[LeafChild] = {
+    val ab = new ArrayBuilder[LeafChild]()
+    keyIterator(key).foreach(ab += _)
+    ab.result()
+  }
+
+  def keyIterator(key: Annotation): Iterator[LeafChild] = new Iterator[LeafChild] {
+    val it = iterateFrom(key)
+    var current: LeafChild = _
+
+    def next(): LeafChild = current
+
+    def hasNext(): Boolean = {
+      it.hasNext && {
+        current = it.next()
+        ordering.equiv(current.key, key)
+      }
     }
   }
 
   def queryByIndex(index: Long): LeafChild = {
     require(index >= 0 && index < nKeys)
-    queryByIndex(index, height - 1, metadata.rootOffset)
+    val node = getLeafNode(index)
+    val localIdx = index - node.firstIndex
+    node.children(localIdx.toInt)
   }
 
+  def queryByInterval(interval: Interval): Iterator[LeafChild] =
+    queryByInterval(interval.start, interval.end, interval.includesStart, interval.includesEnd)
+
+  def queryByInterval(start: Annotation, end: Annotation, includesStart: Boolean, includesEnd: Boolean): Iterator[LeafChild] = {
+    require(Interval.isValid(ordering, start, end, includesStart, includesEnd))
+    val startIdx = if (includesStart) lowerBound(start) else upperBound(start)
+    val endIdx = if (includesEnd) upperBound(end) else lowerBound(end)
+    iterator(startIdx, endIdx)
+  }
+
+  def iterator: Iterator[LeafChild] = iterator(0, nKeys)
+
+  def iterator(start: Long, end: Long) = new Iterator[LeafChild] {
+    assert(start >= 0 && end <= nKeys && start <= end)
+    var pos = start
+    var localPos = 0L
+    var leafNode: LeafNode = _
+
+    def next(): LeafChild = {
```

Comment:
This will happily keep iterating past `end` if the user doesn't check `hasNext` and stop. An `assert(hasNext)` would be more defensive, but maybe this is so performance critical we don't want the extra check.